### ISSUE TITLE
"yodaConditions" shouldn't apply to asymmetric operators

### ISSUE THREAD

**Issue by mattgallagher** [nicklockwood/SwiftFormat#699]
I am forced to turn the "yodaConditions" rule off, due to it breaking code involving asymmetric `==` operator overloads.

Consider the following Swift code:

```swift
try database.fetchOne(User.self, where: .id == myObjectId)
```

With a little operator overloading, this code builds a SELECT statement on a database:

```sql
SELECT * FROM user WHERE id = ? LIMIT 1
```

The `yodaConditions` rule attempts to reorder this `.id == myObjectId` comparison because `.id` is a constant string (a column name). The problem is that the overload of the `==` operator is not symmetric. You can't simply change the order. Signature looks like this:

```swift
func == <Row: DatabaseRow>(_ lhs: Row.ColumnName, _ rhs: DatabaseValueRepresentable) -> DatabaseFilter<Row>
```

Basically, "yodaConditions" should apply only when the left and right sides of the equality operator expect the same type.

(I should probably add that I'm running the latest "develop" branch, bb7d2a3a5701f131902fb4d8ee476c4719a2e4de, although I have no idea is this is a regression.)

**Comment by nicklockwood** [comment]
@mattgallagher unfortunately, since SwiftFormat operates at a token level it cannot know that == has been overloaded, or what the types of its arguments are.

I can't immediately see a good way to prevent it from being applied in this case. Maybe I can add an option to limit it to flow control statements like if/guard/while/where?

**Comment by mattgallagher** [comment]
That might work. But if it's not straightforward, please don't spend too much time on hacking around it. It's a minor issue 😄

**Comment by nicklockwood** [comment]
@mattgallagher I've added a `--yodaswap literals-only` option which will disable swapping `.foo == bar` type conditions, while still applying to cases like `true == foo`. Hopefully that will help in your case.

**Comment by mattgallagher** [comment]
Thanks, @nicklockwood, that's a good solution. Tested and working well!

**Comment by nicklockwood** [comment]
@mattgallagher released as version 0.45.6. Thanks for the report!
